### PR TITLE
[MOB-4239] fix update email with token

### PR DESCRIPTION
**Pull request by roninopf**
## 🔹 Jira Ticket(s)

* [MOB-4239](https://iterable.atlassian.net/browse/MOB-4239)

## ✏️ Description

* fixes functionality to allow setting the JWT for the new email in `updateEmail`

### CODE REVIEW

File: swift-sdk/Internal/AuthManager.swift [hunk 1]
```diff
@@ -53,6 +53,12 @@ class AuthManager: IterableAuthManagerProtocol {
         }
     }
     
+    func setNewTokenAfterUpdateEmail(_ newToken: String) {
```
Comment (tapashmajumder) [inline swift-sdk/Internal/AuthManager.swift:56]: I think we should change the method name to reflect more what it actually does. It appears that this method does not need to be aware that it is called after update email.

Comment (tapashmajumder) [inline swift-sdk/Internal/AuthManager.swift:56]:
```suggestion
    func setNewToken(_ newToken: String) {
```

File: swift-sdk/Internal/InternalIterableAPI.swift [hunk 2]
```diff
@@ -475,14 +486,23 @@ final class InternalIterableAPI: NSObject, PushTrackerProtocol, AuthProvider {
...
         ITBInfo()
         
-        if isEitherUserIdOrEmailSet() && config.authDelegate != nil {
+        if let authToken = authToken {
+            self.authManager.setNewTokenAfterUpdateEmail(authToken)
+        } else if isEitherUserIdOrEmailSet() && config.authDelegate != nil {
             requestNewAuthToken()
-        } else {
-            completeUserLogin()
+            return
```
Comment (tapashmajumder) [inline swift-sdk/Internal/InternalIterableAPI.swift:496]: We should try to exit only in the beginning of method. Exiting in the middle is a code smell.

Comment (tapashmajumder) [reply swift-sdk/Internal/InternalIterableAPI.swift:496]: You can move `completeUserLogin` after line 493 and then add an else clause with `completeUserLogin` added there.

File: swift-sdk/Internal/InternalIterableAPI.swift [hunk 3]
```diff
@@ -475,14 +486,23 @@ final class InternalIterableAPI: NSObject, PushTrackerProtocol, AuthProvider {
         localStorage.userId = _userId
     }
     
-    private func onLogin() {
+    private func onLogin(_ authToken: String? = nil) {
         ITBInfo()
         
-        if isEitherUserIdOrEmailSet() && config.authDelegate != nil {
+        if let authToken = authToken {
```
Comment (tapashmajumder) [inline swift-sdk/Internal/InternalIterableAPI.swift:492]:
```suggestion
        if let authToken = authToken {
           ... setNwAuthToken
           completeUserLogin()
         else if ..... {
            requestNewAuthToken()
         } else {
           completeUserLogin()
         }
```

File: tests/unit-tests/AuthTests.swift [hunk 4]
```diff
@@ -211,6 +211,50 @@ class AuthTests: XCTestCase {
         wait(for: [condition1], timeout: testExpectationTimeout)
     }
     
+    func testUpdateEmailWithTokenParam() {
```
Comment (tapashmajumder) [inline tests/unit-tests/AuthTests.swift:214]: 👏

**Comment by tapashmajumder** [comment]
Logic looks good. I have added some suggestions.

File: tests/unit-tests/AuthTests.swift (hunk 4 above)
Comment (roninopf) [reply tests/unit-tests/AuthTests.swift:214]: :)

**Review by tapashmajumder — APPROVED** [review]
+1 Conditional approval based on doing actual testing. Let me know if you need any help

**Pull request merged**